Implement EncodedVideoChunk

youennf opened the pull request:
#### fb9116e45d33db2050315c4915ad2ed9fe4036fe
<pre>
Implement EncodedVideoChunk
<a href="https://bugs.webkit.org/show_bug.cgi?id=245665">https://bugs.webkit.org/show_bug.cgi?id=245665</a>
rdar://problem/100401981

Reviewed by Eric Carlson.

Add WEB_CODECS compilation guard and runtime flag.
Implement WebCodecsEncodedVideoChunk and expose EncodedVideoChunk
as per <a href="https://w3c.github.io/webcodecs/#encodedvideochunk-interface.">https://w3c.github.io/webcodecs/#encodedvideochunk-interface.</a>

We do not support yet Transferable/Serializable.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/encoded-video-chunk.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/encoded-video-chunk.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/encoded-video-chunk.crossOriginIsolated.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/encoded-video-chunk.crossOriginIsolated.https.any.worker-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.cpp: Added.
(WebCore::WebCodecsEncodedVideoChunk::WebCodecsEncodedVideoChunk):
(WebCore::WebCodecsEncodedVideoChunk::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.h: Added.
(WebCore::WebCodecsEncodedVideoChunk::create):
(WebCore::WebCodecsEncodedVideoChunk::type const):
(WebCore::WebCodecsEncodedVideoChunk::timestamp const):
(WebCore::WebCodecsEncodedVideoChunk::duration const):
(WebCore::WebCodecsEncodedVideoChunk::byteLength const):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.idl: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunkType.h: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunkType.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: <a href="https://commits.webkit.org/254953@main">https://commits.webkit.org/254953@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b3504f6f95850d3fd190ace7a604ad3114615a1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35389 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21373 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100103 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94821 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33888 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83155 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/96498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96468 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77615 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26803 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81751 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69832 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82327 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34975 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/15525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/77324 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32776 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16506 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26666 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3456 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36552 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/79916 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38477 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35595 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17508 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.idl:
```diff
@@ -0,0 +1,51 @@
+  undefined copyTo([AllowShared] BufferSource destination);
+};
+
+typedef [EnforceRange] long long WebCodecsEncodedVideoChunkInitTimestamp;
+typedef [EnforceRange] unsigned long long WebCodecsEncodedVideoChunkInitDuration;
```
Why not use these for `WebCodecsEncodedVideoChunk. timestamp` and `WebCodecsEncodedVideoChunk. duration` as well?

youennf replied:
This aligns with the spec.
We want to enforce range when creating the video chunk from a dictionary.
The video chunk attribute getters should be fine since they will expose already validated values.

eric-carlson reviewed (approved):
r=me

youennf committed:
Implement EncodedVideoChunk
https://bugs.webkit.org/show_bug.cgi?id=245665
rdar://problem/100401981

Reviewed by Eric Carlson.

Add WEB_CODECS compilation guard and runtime flag.
Implement WebCodecsEncodedVideoChunk and expose EncodedVideoChunk
as per https://w3c.github.io/webcodecs/#encodedvideochunk-interface.

We do not support yet Transferable/Serializable.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/encoded-video-chunk.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/encoded-video-chunk.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/encoded-video-chunk.crossOriginIsolated.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/encoded-video-chunk.crossOriginIsolated.https.any.worker-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.cpp: Added.
(WebCore::WebCodecsEncodedVideoChunk::WebCodecsEncodedVideoChunk):
(WebCore::WebCodecsEncodedVideoChunk::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.h: Added.
(WebCore::WebCodecsEncodedVideoChunk::create):
(WebCore::WebCodecsEncodedVideoChunk::type const):
(WebCore::WebCodecsEncodedVideoChunk::timestamp const):
(WebCore::WebCodecsEncodedVideoChunk::duration const):
(WebCore::WebCodecsEncodedVideoChunk::byteLength const):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.idl: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunkType.h: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunkType.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: https://commits.webkit.org/254953@main
